Hl7 10 am signup endpoint

amathew8 opened the pull request:
### Related JIRA tickets:
>Please enter the whole URL so we can just click/copy it
-https://jira.amida.com/browse/HL7-10

### What exactly does this PR do?
- Created a signup endpoint for users to create account
- Added checks for for duplicate emails, and usernames
-  Added conditions for email and password


### Do ANY of these changes introduce a breaking change? (in-scope or otherwise)

- No

### Manual test cases?
- First enter username, email, password in Postman and send post request to api/users/signup
- After creating user, try sending again to see error message
- Also try to use incorrect emails or passwords that are to short (<6 char)

### Any additional context/background?

- The user.test will only work the first time for now as once you create a user it cannot be created again

### Screenshots (if appropriate):

amathew8 committed:
Created signup endpoint and modified tests

amathew8 committed:
Merge branch 'develop' of https://github.com/amida-tech/hl7-telescoper-api-server into hl7-10-AM-signup-endpoint

amathew8 committed:
Created signup endpoint

Perry5 commented on server/user/user.controller.js:
```diff
@@ -23,20 +24,58 @@ function get(req, res) {
+    $or: [
+      { email: req.body.email },
+      { username: req.body.username }
+      // eslint-disable-next-line consistent-return
```
Is there a way to work around the linting error here? instead of disabling it?

orndorffgrant replied:
yeah let's remove all of these eslint-disable lines - generally we shouldn't be disabling linting rules unless we have a strong reason

Perry5 commented on server/user/user.controller.js:
```diff
@@ -23,20 +24,58 @@ function get(req, res) {
+    // eslint-disable-next-line eqeqeq
+    if (docs.length >= 1) {
+      if (docs[0].email === req.body.email) {
+        return res.status(409)
```
Can we use the `httpStatus` as well as the `APIError` modules here? see how its done in `server/hl7/hl7.controller.js` for an example.

orndorffgrant replied:
Agree, nice use of correct http status codes though! https://http.cat/409

Perry5 commented on server/user/user.controller.js:
```diff
@@ -23,20 +24,58 @@ function get(req, res) {
+          });
+        // eslint-disable-next-line eqeqeq
+      } else if (docs[0].username === req.body.username) {
+        return res.status(409)
```
See comment above

Perry5 commented on server/user/user.controller.js:
```diff
@@ -23,20 +24,58 @@ function get(req, res) {
+    // eslint-disable-next-line consistent-return
+    bcrypt.hash(req.body.password, 10, (error, hash) => {
+      if (error) {
+        return res.status(500)
```
See comment above

Perry5 commented on server/user/user.controller.js:
```diff
@@ -23,20 +24,58 @@ function get(req, res) {
+      });
+      // Save user
+      user.save()
+        .then(savedUser => res.json(savedUser))
```
We should probably return a `201 Created` here since we are creating a resource for the first time. This currently will return a `200 OK`. See above comments on how to use the `httpStatus` library for this.

Perry5 commented on server/user/user.model.js:
```diff
@@ -11,10 +11,17 @@ const UserSchema = new mongoose.Schema({
+  password: {
+    type: String,
+    required: true,
+
```
Delete empty spaces please

Perry5 commented on server/user/user.route.js:
```diff
@@ -7,10 +7,10 @@ const router = express.Router(); // eslint-disable-line new-cap
-  /** POST /api/users - Create new user */
-  .post(validate(paramValidation.createUser), userCtrl.create);
+  .get(userCtrl.list);
+  //
```
Delete unused comments please

Perry5 commented on server/user/user.test.js:
```diff
@@ -20,103 +21,115 @@ after((done) => {
   let user = {
-    username: 'KK123',
-    mobileNumber: '1234567890'
+    username: 'fakeName',
```
There is nothing wrong with this user object, but the work "Fake" just sounds weird lol. Maybe call it "TestUser"? Not a big deal, a matter of preference.

orndorffgrant replied:
lol 🤷‍♂

Perry5 commented on server/user/user.test.js:
```diff
@@ -20,103 +21,115 @@ after((done) => {
-        })
-        .catch(done);
-    });
+/*
```
Delete any unused code please. Unless you know for sure it will be used later.

orndorffgrant replied:
looks like most of this should be uncommented and left in actually

orndorffgrant commented on config/param-validation.js:
```diff
@@ -5,7 +5,9 @@ module.exports = {
       username: Joi.string().required(),
-      mobileNumber: Joi.string().regex(/^[1-9][0-9]{9}$/).required()
+      // eslint-disable-next-line max-len,no-useless-escape
+      email: Joi.string().regex(/^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/).required(),
```
Joi actually has a built in email function so we don't need this huge regex
```suggestion
      email: Joi.string().email({ minDomainSegments: 2 }).required(),
```
and then we won't need the `eslint-disable` comment either


Aside: here's some fun reading on email regexes: https://www.regular-expressions.info/email.html

orndorffgrant commented on data/hl7-uploads/500HL7Messages.txt:
```diff
@@ -0,0 +1,5366 @@
+MSH|^~\&|App3|Facility2|App4|Facility6|201906191028||OMP^O09|3926758|D^I|2.2
```
please delete this file :)

Perry5 reviewed (commented):
Good work here! Functionality seems to work as expected/described. I left some in line comments as well as these.

- Seems like you committed the file in `data/hl7-uploads`. I don't think we want to commit this file. Everything in `/data` should be ignored. Make sure you `.gitignore` file is up to date. 
- See Grant's comment on my PR here: https://github.com/amida-tech/hl7-telescoper-api-server/pull/2#discussion_r297245245 - "Run yarn config set save-prefix "" so that yarn won't add false before version numbers"
- Might be a good idea to add a test case for trying to create a user that already exist.

orndorffgrant commented on package.json:
```diff
@@ -40,6 +40,7 @@
     "boilerplate"
   ],
   "dependencies": {
+    "bcrypt": "false3.0.6",
```
```suggestion
    "bcrypt": "3.0.6",
```
and run `yarn config set save-prefix ""` so that yarn won't add `false` before version numbers in the future

orndorffgrant commented on server/user/user.controller.js:
```diff
@@ -23,20 +24,58 @@ function get(req, res) {
-  const user = new User({
-    username: req.body.username,
-    mobileNumber: req.body.mobileNumber
+  // First check if username or email already in database
```
There's a race condition in this function. Let's work together to fix it before we merge this.

orndorffgrant commented on server/user/user.test.js:
```diff
@@ -20,103 +21,115 @@ after((done) => {
+          expect(res.body.email)
+            .to
+            .equal(user.email);
+          // eslint-disable-next-line no-shadow
```
lets fix this linting error instead of disabling it

orndorffgrant reviewed (changes requested):
Overall looks good! Some small things, mostly stylistic - one functional problem

orndorffgrant commented:
Also, @Perry5 I don't think we need to change the endpoint url from POST `/api/users` to POST `/api/users/signup` - thoughts?

amathew8 committed:
Fixed ESLint surpress warning, possible race condition, and added API errors

amathew8 committed:
Deleted 500 Hl7 Message file

amathew8 committed:
Fixed eslint error

amathew8 commented:
-Fixed comments, ESLint surprise warnings, and race condition problems. 
-Added API errors 
-Deleted HL7 Sample messages file
-Changed fakeName to testName for all fields

amathew8 committed:
Fixed error where users were being created even after error showed

Perry5 commented:
@orndorffgrant I agree- I don't think there is really a need to change the url. We should probably keep it as `api/users/` actually.

Perry5 commented on data/hl7-sample/test.txt:
```diff
@@ -0,0 +1 @@
+this is a test text file
```
Can we delete this file? @amathew8 make sure you don't add/commit files that get added to the `/data` folder if you don't want them to show up on github.

orndorffgrant replied:
I changed the `.gitignore` to only ignore `/data/hl7-uploads` so that we could keep test files in `/data/hl7-sample` and this file is being used in the tests to test the upload endpoint.

amathew8 committed:
Fixed final lint surpresses and changed url to api/users

orndorffgrant reviewed (approved):
Thanks for the updates!
